src/transformers/models/jukebox/modeling_jukebox.py

```diff
@@ -0,0 +1,3228 @@
+                # change from straight-through estimator
+                x_quantised = x_quantised.detach()
+            xs_quantised.append(x_quantised)
+            commit_losses.append(commit_loss)
+            if self.training:
+                metrics.append(metric)
+        return zs, xs_quantised, commit_losses, metrics
+
+
+# TODO replace FFT calls with torch.fft
+def stft(sig, hps):
+    return torch.stft(
+        sig,
+        hps.n_fft,
+        hps.hop_length,
+        win_length=hps.window_size,
+        window=torch.hann_window(hps.window_size, device=sig.device),
+    )
+
+
+# TODO replace spec
+def spec(x, hps):
+    return torch.norm(stft(x, hps), p=2, dim=-1)
+
+
+# TODO check if can be removed
+
+
+class DefaultSTFTValues:
+    def __init__(self, hps):
+        self.sr = hps.sr
+        self.n_fft = 2048
+        self.hop_length = 256
+        self.window_size = 6 * self.hop_length
+
+
+def norm(x):
+    return (x.view(x.shape[0], -1) ** 2).sum(dim=-1).sqrt()
+
+
+def squeeze(x):
+    if len(x.shape) == 3:
+        assert x.shape[-1] in [1, 2]
+        x = torch.mean(x, -1)
+    if len(x.shape) != 2:
+        raise ValueError(f"Unknown input shape {x.shape}")
+    return x
+
+
+def spectral_loss(x_in, x_out, hps):
+    hps = DefaultSTFTValues(hps)
+    spec_in = spec(squeeze(x_in.float()), hps)
+    spec_out = spec(squeeze(x_out.float()), hps)
+    return norm(spec_in - spec_out)
+
+
+def spectral_convergence(x_in, x_out, hps, epsilon=2e-3):
+    hps = DefaultSTFTValues(hps)
+    spec_in = spec(squeeze(x_in.float()), hps)
+    spec_out = spec(squeeze(x_out.float()), hps)
+
+    gt_norm = norm(spec_in)
+    residual_norm = norm(spec_in - spec_out)
+    mask = (gt_norm > epsilon).float()
+    return (residual_norm * mask) / torch.clamp(gt_norm, min=epsilon)
+
+
+class STFTValues:
+    def __init__(self, hps, n_fft, hop_length, window_size):
+        self.sr = hps.sr
+        self.n_fft = n_fft
+        self.hop_length = hop_length
+        self.window_size = window_size
+
+
+def multispectral_loss(x_in, x_out, hps):
+    losses = []
+    assert len(hps.multispec_loss_n_fft) == len(hps.multispec_loss_hop_length) == len(hps.multispec_loss_window_size)
+    args = [hps.multispec_loss_n_fft, hps.multispec_loss_hop_length, hps.multispec_loss_window_size]
+    for n_fft, hop_length, window_size in zip(*args):
+        hps = STFTValues(hps, n_fft, hop_length, window_size)
+        spec_in = spec(squeeze(x_in.float()), hps)
+        spec_out = spec(squeeze(x_out.float()), hps)
+        losses.append(norm(spec_in - spec_out))
+    return sum(losses) / len(losses)
+
+
+def average_metrics(_metrics):
+    metrics = {}
+    for _metric in _metrics:
+        for key, val in _metric.items():
+            if key not in metrics:
+                metrics[key] = []
+            metrics[key].append(val)
+    return {key: sum(vals) / len(vals) for key, vals in metrics.items()}
+
+
+class VQVAE(nn.Module):
```

Comment:
@ArthurZucker - can we make this a `PreTrainedModel`?

Reply:
-> and name it `JukeboxVQVAE`

Reply:
Yes :)